src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -817,38 +738,37 @@ public Completable invoke(String method, Object... args) {
             return observable.doOnDispose(() -> {
                 if (subscriptionCount.decrementAndGet() == 0) {
                     CancelInvocationMessage cancelInvocationMessage = new CancelInvocationMessage(null, invocationId);
-                    sendHubMessage(cancelInvocationMessage);
-                    if (connectionState != null) {
-                        connectionState.tryRemoveInvocation(invocationId);
+                    sendHubMessageWithLock(cancelInvocationMessage);
+                    ConnectionState connectionStateInner = this.state.getConnectionState();
```

Comment:
In situations like these, I think it's better to use the connectionState from the closure in case it's changed since the lock was released.